nomos-mix/network/src/behaviour.rs

```diff
@@ -115,82 +150,86 @@ where
         Ok(())
     }
 
-    fn add_negotiated_peer(&mut self, peer_id: PeerId, connection_id: ConnectionId) -> bool {
-        tracing::debug!(
-            "Adding to connected_peers: peer_id:{:?}, connection_id:{:?}",
-            peer_id,
-            connection_id
-        );
-        self.negotiated_peers
-            .entry(peer_id)
-            .or_default()
-            .insert(connection_id)
-    }
-
-    fn remove_negotiated_peer(&mut self, peer_id: &PeerId, connection_id: &ConnectionId) {
-        if let Some(connections) = self.negotiated_peers.get_mut(peer_id) {
-            tracing::debug!(
-                "Removing from connected_peers: peer:{:?}, connection_id:{:?}",
-                peer_id,
-                connection_id
-            );
-            connections.remove(connection_id);
-            if connections.is_empty() {
-                self.negotiated_peers.remove(peer_id);
-            }
-        }
-    }
-
-    /// SHA-256 hash of the message
     fn message_id(message: &[u8]) -> Vec<u8> {
         let mut hasher = Sha256::new();
         hasher.update(message);
         hasher.finalize().to_vec()
     }
 
+    fn run_conn_maintenance(&mut self) {
+        let (peers_to_close, peers_to_connect) = self.conn_maintenance.reset();
+
+        // Schedule events to close connections
+        peers_to_close.into_iter().for_each(|addr| {
+            if let Some(peer_id) = self.peer_address_map.peer_id(&addr) {
+                self.events.push_back(ToSwarm::CloseConnection {
+                    peer_id,
+                    connection: CloseConnection::All,
+                });
+            }
+        });
+        // Schedule events to connect to peers
+        peers_to_connect.into_iter().for_each(|addr| {
+            self.events.push_back(ToSwarm::Dial {
+                opts: DialOpts::from(addr),
+            });
+        });
+    }
+
```

Comment:
Yes! this is exactly what I mean. I think this is a better way to do this. Thanks!